In [1]:
# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Install mlxtend library
!pip install mlxtend

In [3]:
# Import functions for association rule mining
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

In [4]:
# Importing file module and uploading the data file to colab environment
from google.colab import files
uploaded=files.upload()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Saving Online retail.xlsx to Online retail (1).xlsx


In [5]:
# Read the Excel file
df=pd.read_excel('Online retail.xlsx',header=None)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [6]:
# Display first few rows of the dataframe
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

In [7]:
# Displays the last few rows of the dataframe
df.tail()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

In [8]:
# Displays the size of the dataframe
df.size

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


7501

In [9]:
# Displays the shape of the dataframe
df.shape

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(7501, 1)

In [10]:
# Displays the information of the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [11]:
# Displays the statistical values
df.describe()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0
count      7501
unique     5176
top     cookies
freq        223

In [12]:
# Displays the no of missing values
df.isnull().sum()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0    0
dtype: int64

In [13]:
# Displays the sum of duplicated values
df.duplicated().sum()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


2325

In [14]:
# Split the single column into lists of items
df['Transactions'] = df.iloc[:, 0].str.split(',')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [15]:
# Verify the first few rows of the 'Transactions' column
df['Transactions'].head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0    [shrimp, almonds, avocado, vegetables mix, gre...
1                           [burgers, meatballs, eggs]
2                                            [chutney]
3                                    [turkey, avocado]
4    [mineral water, milk, energy bar, whole wheat ...
Name: Transactions, dtype: object

In [16]:
# Convert the list of transactions into a one-hot encoded dataframe
transactions = df['Transactions'].tolist()
te = TransactionEncoder()
te_array = te.fit(transactions).transform(transactions)
encoded_df = pd.DataFrame(te_array, columns=te.columns_)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [17]:
encoded_df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
7496       False    False              False      False    False        False   
7497       False    False              False      False    False        False   
7498       False    False              False      False    False        False   
7499       False    False              False      False    False        False   
7500       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
7496  False           False      False        False  ...   False   
7497  False           False      False        False  ...   False   
7498  False           False      False        False  ...   False   
7499  False           False      False        False  ...   False   
7500  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   
7500           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False        False     False  
7500              False             False  False         True     False  

[7501 rows x 120 columns]

In [18]:
# Applying apriori algorithm
frequent_itemsets = apriori(encoded_df,min_support = 0.05,use_colnames=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [19]:
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                    itemsets
0   0.087188                   (burgers)
1   0.081056                      (cake)
2   0.059992                   (chicken)
3   0.163845                 (chocolate)
4   0.080389                   (cookies)
5   0.051060               (cooking oil)
6   0.179709                      (eggs)
7   0.079323                  (escalope)
8   0.170911              (french fries)
9   0.063325           (frozen smoothie)
10  0.095321         (frozen vegetables)
11  0.052393             (grated cheese)
12  0.132116                 (green tea)
13  0.098254               (ground beef)
14  0.076523            (low fat yogurt)
15  0.129583                      (milk)
16  0.238368             (mineral water)
17  0.065858                 (olive oil)
18  0.095054                  (pancakes)
19  0.071457                    (shrimp)
20  0.050527                      (soup)
21  0.174110                 (spaghetti)
22  0.068391                  (tomatoes)
23  0.062525                    (turkey)
24  0.058526          (whole wheat rice)
25  0.052660  (chocolate, mineral water)
26  0.050927       (eggs, mineral water)
27  0.059725  (spaghetti, mineral water)

In [20]:
# Generate association rules
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.001,num_itemsets=len(frequent_itemsets))

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  consequent support  \
0      (chocolate)  (mineral water)            0.163845            0.238368   
1  (mineral water)      (chocolate)            0.238368            0.163845   
2           (eggs)  (mineral water)            0.179709            0.238368   
3  (mineral water)           (eggs)            0.238368            0.179709   
4      (spaghetti)  (mineral water)            0.174110            0.238368   
5  (mineral water)      (spaghetti)            0.238368            0.174110   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.052660    0.321400  1.348332               1.0  0.013604    1.122357   
1  0.052660    0.220917  1.348332               1.0  0.013604    1.073256   
2  0.050927    0.283383  1.188845               1.0  0.008090    1.062815   
3  0.050927    0.213647  1.188845               1.0  0.008090    1.043158   
4  0.059725    0.343032  1.439085               1.0  0.018223    1.159314   
5  0.059725    0.250559  1.439085               1.0  0.018223    1.102008   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.308965  0.150648   0.109018    0.271158  
1       0.339197  0.150648   0.068256    0.271158  
2       0.193648  0.138707   0.059103    0.248515  
3       0.208562  0.138707   0.041372    0.248515  
4       0.369437  0.169312   0.137421    0.296796  
5       0.400606  0.169312   0.092566    0.296796


**Analysis and Interpretation**

1. High-Lift Rules Indicating Strong Relationships
Rule:

{
spaghetti
}
→
{
mineral water
}
{spaghetti}→{mineral water}

Lift: 1.439

Support: 0.059 (5.9%)

Confidence: 0.343 (34.3%)

Interpretation: Customers who buy spaghetti are 34.3% likely to also buy mineral water, and this is 43.9% higher than random chance. Spaghetti and mineral water are complementary items, often purchased together.

Insight: Consider bundling these products or running joint promotions.


2. Cross-Category Relationships
Rule:

{
chocolate
}
→
{
mineral water
}
{chocolate}→{mineral water}

Lift: 1.348

Support: 0.052 (5.2%)

Confidence: 0.321 (32.1%)

Interpretation: Chocolate buyers are 32.1% likely to purchase mineral water, and the relationship is 34.8% stronger than expected by chance.

Insight: This suggests a potential pattern of indulgence offset with healthier choices. Promotions pairing chocolate and beverages like mineral water could be effective.

3. Eggs and Mineral Water
Rule:

{
eggs
}
→
{
mineral water
}
{eggs}→{mineral water}

Lift: 1.188

Support: 0.051 (5.1%)

Confidence: 0.283 (28.3%)

Interpretation: Egg buyers are 28.3% likely to purchase mineral water, but the lift is moderate (18.8% above chance). This suggests a weaker but notable relationship.

Insight: Highlight recipes or promotions involving eggs and mineral water in newsletters or ads.


4. Bidirectional Patterns

Many rules show bidirectional relationships:

{
chocolate
}
→
{
mineral water
}
{chocolate}→{mineral water} and
{
mineral water
}
→
{
chocolate
}
{mineral water}→{chocolate}

These rules suggest items are frequently co-purchased, regardless of the starting point.

Insight: Such products are likely to be co-promoted effectively.

5. Focus on High Support Rules
High support rules like
{
spaghetti
→
mineral water
}
{spaghetti→mineral water} (5.9%) represent patterns that occur in a substantial proportion of transactions.

Insight: High-support, high-confidence rules can guide inventory management to ensure consistent stock availability.


**General Insights on Customer Behavior**

1. Pairing Essentials with Treats:

Relationships like
chocolate
↔
mineral water
chocolate↔mineral water suggest customers balance indulgent purchases with essentials or healthier options.

2. Recipe-Driven Shopping:

Combinations like
spaghetti
→
mineral water
spaghetti→mineral water may indicate customers buying items for specific meals.

3. Mineral Water as a Key Driver:

Mineral water appears frequently in the consequent of rules, indicating its role as a staple product that drives additional purchases.

**Interview Questions**

1. What is lift and why is it important in Association rules?

Lift is a measure of how much more likely the consequent (item Y) is purchased when the antecedent (item X) is purchased, compared to its usual purchase rate. In other words, it quantifies the strength of association between two items.

Lift > 1: Indicates a positive association between the antecedent and consequent, meaning they are more likely to occur together than by chance.

Lift = 1: Implies no association; the items occur together as often as expected by random chance.

Lift < 1: Suggests a negative association, meaning the antecedent and consequent are less likely to occur together.

Importance:
* Identifying true relationships
* Prioritizing rules
* Avoiding misleading insights

2. What is support and Confidence. How do you calculate them?


**Support:** The proportion of transactions that contain both the antecedent (item X) and the consequent (item Y). It measures the frequency of the itemset (X and Y together) in the dataset.

Support(X -> Y) = (Number of transactions containing both X and Y) / (Total number of transactions)


**Confidence:** The probability of purchasing the consequent (item Y) given that the antecedent (item X) has been purchased. It measures the strength of the implication.

Confidence(X -> Y) = (Number of transactions containing both X and Y) / (Number of transactions containing X)

3. What are some limitations or challenges of Association rules mining?



*   Finding meaningful rules

*   Handling rare items

*   Spurious association
*   Computational complexity


*   Data sparsity


*   Overfitting



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
